### PR TITLE
Reference interpreter

### PR DESCRIPTION
**Pull request by dhil**
This patch adds a reference interpreter for the stack switching proposal. It implements all of the proposed instructions.

The patch also includes many updates from WebAssembly/spec, WebAssembly/function-references, WebAssembly/exception-handling, and WebAssembly/gc.

The contents of this patch are comprised of multiple smaller patches which have already been reviewed by @rossberg -- though, if you are keen please take a look.

### CODE REVIEW

File: interpreter/syntax/ast.ml [hunk 1]
```diff
@@ -166,8 +167,9 @@ and instr' =
   | ContNew of idx                    (* create continuation *)
   | ContBind of idx * idx             (* bind continuation arguments *)
   | Suspend of idx                    (* suspend continuation *)
-  | Resume of idx * (idx * idx) list  (* resume continuation *)
-  | ResumeThrow of idx * idx * (idx * idx) list (* abort continuation *)
+  | Resume of idx * (idx * hdl) list  (* resume continuation *)
+  | ResumeThrow of idx * idx * (idx * hdl) list (* abort continuation *)
+  | Switch of idx * idx               (* direct switch continuation *)
```
Comment (Alan-Liang) [inline interpreter/syntax/ast.ml:172]: Should we have `switch $ct1 $ct2 $e` (i.e., `Switch of idx * idx * idx`) here?

Comment (dhil) [reply interpreter/syntax/ast.ml:172]: Not anymore, c.f. #76.

Comment (Alan-Liang) [reply interpreter/syntax/ast.ml:172]: Oh, I missed that.

**Comment by dhil** [comment]
Following on from the discussion we had Monday, I've removed the `barrier` instruction from the reference interpreter.

**Comment by dhil** [comment]
This is ready for review / to be merged once #90 has been merged.

**Comment by dhil** [comment]
@tlively and @slindley this patch is ready for review now. PTAL. The patch includes some updates to the legacy proposal documents. I think these should be consolidated under stack-switching/ along side the legacy bag-o-stacks documents (if we decide to keep any of these documents around). But I think that is for another patch.

**Review by tlively — APPROVED** [review]
🎉 

(mostly rubberstamp)

**Review by slindley — APPROVED** [review]
LGTM

**Pull request merged**